usr/man/man/usr/share/man/man5/audit_remote.5
'\" te
.\"  Copyright (c) 2009, 2010, Oracle and/or its affiliates. All rights reserved.
.TH audit_remote 5 "22 Jun 2010" "SunOS 5.11" "Standards, Environments, and Macros"
.SH NAME
audit_remote \- send Solaris audit logs to a remote server
.SH SYNOPSIS
.LP
.nf
\fB/usr/lib/security/audit_remote.so\fR
.fi

.SH DESCRIPTION
.sp
.LP
The \fBaudit_remote\fR plugin module for Solaris audit, \fB/usr/lib/security/audit_remote.so\fR, sends binary audit records (\fBaudit.log\fR(4)) to audit servers as they are configured with \fBauditconfig\fR(1M).
.sp
.LP
The \fBaudit_remote\fR plugin is loaded by \fBauditd\fR(1M) if the plugin is configured as an active via \fBauditconfig\fR. Use the \fBauditconfig\fR \fB-setplugin\fR option to change all the plugin related configuration parameters.
.SS "Object Attributes"
.sp
.LP
The following attributes specify the configuration of \fBaudit_remote\fR plugin:
.sp
.ne 2
.mk
.na
\fB\fBp_hosts\fR\fR
.ad
.sp .6
.RS 4n
.sp
.in +2
.nf
\fIhost1\fR[:[\fIport1\fR][:\fImech1\fR]][,\fIhost2\fR[:[\fIport2\fR][:\fImech2\fR]],... \e
    \fIhostn\fR[:[\fIportn\fR][:\fImechn\fR]]]
.fi
.in -2
.sp

A list of audit hosts/servers. Audit records are sent to the first available host. If a host is unreachable or a timeout occurs while sending data, the next host in the list is tried. If connection to all hosts fails, the list is tried again from the beginning. 
.sp
The \fIhost\fR part of a \fBp_hosts\fR entry can be in any form acceptable to \fBgetipnodebyname\fR(3SOCKET).
.sp
The \fIport\fR part of a \fBp_hosts\fR entry is the port on host that is contacted to initiate an audit server connection. If not specified, the port number is that assigned to the \fBsolaris-audit\fR service. See \fBgetservbyname\fR(3XNET).
.sp
The \fBmech\fR part of a \fBp_host\fR entry is the GSS-API mechanism name (\fBmech\fR(4)). If not specified, the local host's default mechanism is used. The recommended mechanism is \fBkerberos_v5\fR.
.RE

.sp
.ne 2
.mk
.na
\fB\fBp_retries\fR\fR
.ad
.sp .6
.RS 4n
The number of retries for connecting to and sending data to a server. 
.sp
The default value is \fB3\fR.
.RE

.sp
.ne 2
.mk
.na
\fB\fBp_timeout\fR\fR
.ad
.sp .6
.RS 4n
The number of seconds in which a connection/sending data timeouts. 
.sp
The default value is \fB5\fR seconds.
.RE

.sp
.ne 2
.mk
.na
\fB\fBqsize\fR\fR
.ad
.sp .6
.RS 4n
The maximum number of outstanding audit records to keep.
.sp
The default is the value of the kernel queue control high water mark. See \fBauditconfig\fR(1M).
.RE

.sp
.LP
If set to 0, the default value is the value of the kernel queue control high water mark. See \fBauditconfig\fR(1M).
.SS "GSS SESSION"
.sp
.LP
The \fBaudit_remote plugin\fR is a TCP client that authenticates configured audit servers using the GSS-API (\fBlibgss\fR(3LIB)). Binary Solaris Audit records are sent with integrity and confidentiality protection as per-message tokens generated by \fBgss_wrap\fR(3GSS).
.sp
.LP
The plugin initiates a TCP connection to an audit server (\fIhost:port:mech\fR) and establishes a GSS security context (with \fBgss_init_sec_context\fR(3GSS)), with appropriate security mechanism (\fBmech\fR(4)). 
.sp
.LP
If no port is specified, the service name \fBsolaris-audit\fR is looked up to obtain a TCP port number. If no mechanism is specified, the \fBGSS_C_NO_OID\fR is used as a \fBmech_type\fR parameter of \fBgss_init_sec_context\fR(3GSS), and causes the underlying \fBGSS-API\fR to use the local default mechanism. 
.sp
.LP
\fBgss_init_sec_context\fR(3GSS) uses \fBGSS_C_NO_CREDENTIAL\fR as the initiator credential handle and a target name of the form \fBaudit@<host_fqdn>\fR. The server is expected to use \fBgss_accept_sec_context\fR(3GSS) to complete the context establishment. 
.sp
.LP
Once the security context is established, the client (\fBaudit_remote\fR plugin) calls \fBgss_wrap\fR(3GSS) to achieve the confidentiality of the transferred payload - the audit records. The server is expected to use \fBgss_unwrap\fR(3GSS) to unwrap the received data and \fBgss_get_mic\fR(3GSS) to obtain the MIC (Message Integrity Code) to be later sent back to the plugin as a message retrieval acknowledgment.
.sp
.LP
For example, if the \fBkerberos_v5\fR mechanism is configured as \fBGSS_API\fR mechanism on the client and both sides agree on using this mechanism, the client side has to be eligible to non-interactively gain session keys for the \fBaudit/<host_fqdn>@<REALM>\fR principal from the Kerberos  KDC/TGS. At the same time the identity running the audit server application has to have the long term keys associated with the \fBaudit/<host_fqdn>@<REALM>\fR principal stored in the \fBkeytab\fR file (\fBkrb5.conf\fR(4)) to be able to decrypt the session keys.
.sp
.LP
The \fBaudit_remote\fR plugin initiates a connection to first server in the \fBp_hosts\fR list. If the connection fails or audit record sends are not responded to in \fBp_timeout\fR seconds, after \fBp_retries\fR attempts the plugin tries to connect to the next server. If the connection to the last server fails, the plugin retries to connect to the first host in the list. \fBaudit_warn\fR(1M) is executed at every unsuccessful attempt to connect to the server or send timeout with the plugin option plugin \fBaudit_remote.so retry <count> <error>.<error>\fR is connection \fB<host:port> <the network error>\fR\&. An \fBEPROTO\fR network error indicates that the client plugin did not get a successful protocol version handshake.
.SS "PROTOCOL DESCRIPTION"
.sp
.LP
All protocol messages are preceded by the 4 octets of the size of the data to follow. This size is in network byte order.
.sp
.LP
The protocol begins with version negotiation followed by a \fBGSS-API\fR security context token exchange. On error the connection is closed (and any output token optionally sent).
.sp
.LP
The version negotiation takes place in the clear with the plugin sending an octet array of the comma (\fB,\fR) separated list of versions supported. The current version number is the characters \fB01\fR. The receiver is expected to respond with the version that they accept (in the current case that is the characters \fB01\fR). A mismatch is considered an error and the connection is closed.
.sp
.LP
The version octet array sent by the plugin and the version characters accepted by the receiver are concatenated together to make up the application data field of the channel bindings of the GSS security context establishment.
.sp
.in +2
.nf
<plugin version characters> || <server accepted version characters>
"||" represents concatenation
.fi
.in -2

.sp
.LP
Subsequent tokens contain a 64 bit sequence number in network byte order and a single audit record (\fBaudit.log\fR(4)); the client uses confidentiality protection. wrap (64 bit sequence number || audit record)
.sp
.LP
The server acknowledges the receipt (and is then responsible for any data loss) with the received 64 bit sequence number and a MIC token of the unwrapped 64 bit sequence number and audit record. MIC verification on the client side acknowledges the audit record can be freed and not saved for possible retransmission. 
.sp
.in +2
.nf
64 bit sequence number || mic (64 bit sequence number || audit record)
.fi
.in -2

.sp
.LP
Secure remote audit client/server communication flow:
.sp
.in +2
.nf
1) Client <--> Server - TCP handshake

2) Client <--> Server - protocol version negotiation:
   a) Client  --> Server - send data size - uint32_t value (2)
   b) Client  --> Server - send clear text message of the versions
                           supported comma separated, e.g.,
                           "01,02,03" for versions 1 and 2 and 3.
                           The only version supported at present is
                           "01"
   c) Client <--  Server - send data size - uint32_t value (2)
   d) Client <--  Server - send clear text version selected
                           ("01")
   :no version match; close connection; try next host

3) Security context initiation:
   a) Client - Construct channel bindings:
         initiator address type (GSS_C_AF_NULLADDR)
         acceptor address type (GSS_C_AF_NULLADDR)
         application data value (4 octets  "0101")
   b) Client  --> Server - send token (data) size - uint32_t value
   c) Client  --> Server - GSS-API per-context token
   d) Client <--  Server - send token (data) size
   e) Client <--  Server - GSS-API per-context token
      :repeat a-e until security context is initialized; if unsuccessful,
      close connection; try next host

4) Client - transmit thread, when audit record to be sent:
   a) Client  --> Server - send data size
   b) Client  --> Server - GSS-API per-message token
                  wrap (sequence number || audit record)
      :repeat a-b while less than max (qsize) outstanding records

 5) Client - receive thread:
    a) Client <--  Server - receive data size - uint32_t value
    b) Client <--  Server - receive sequence number - uint64_t value
    c) Client <--  Server - receive MIC
    d) Client             - MIC verification - OK

usr/src/lib/auditd_plugins/remote/audit_remote.c
/*
 * CDDL HEADER START
 *
 * The contents of this file are subject to the terms of the
 * Common Development and Distribution License (the "License").
 * You may not use this file except in compliance with the License.
 *
 * You can obtain a copy of the license at usr/src/OPENSOLARIS.LICENSE
 * or http://www.opensolaris.org/os/licensing.
 * See the License for the specific language governing permissions
 * and limitations under the License.
 *
 * When distributing Covered Code, include this CDDL HEADER in each
 * file and include the License file at usr/src/OPENSOLARIS.LICENSE.
 * If applicable, add the following below this CDDL HEADER, with the
 * fields enclosed by brackets "[]" replaced with your own identifying
 * information: Portions Copyright [yyyy] [name of copyright owner]
 *
 * CDDL HEADER END
 */
/*
 * Copyright (c) 2009, 2011, Oracle and/or its affiliates. All rights reserved.
 *
 * send audit records to remote host
 *
 */

/*
 * auditd_plugin_open(), auditd_plugin() and auditd_plugin_close()
 * implement a replaceable library for use by auditd; they are a
 * project private interface and may change without notice.
 */

#include <arpa/inet.h>
#include <assert.h>
#include <audit_plugin.h>
#include <bsm/audit.h>
#include <bsm/audit_record.h>
#include <bsm/libbsm.h>
#include <errno.h>
#include <fcntl.h>
#include <gssapi/gssapi.h>
#include <libintl.h>
#include <netdb.h>
#include <pthread.h>
#include <rpc/rpcsec_gss.h>
#include <secdb.h>
#include <signal.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <strings.h>
#include <ctype.h>
#include <sys/param.h>
#include <sys/socket.h>
#include <sys/types.h>
#include <unistd.h>
#include <poll.h>

#include "audit_remote.h"

#define	DEFAULT_TIMEOUT	5	/* default connection timeout (in secs) */
#define	NOSUCCESS_DELAY	20	/* unsuccessful delivery to all p_hosts */

#define	FL_SET		B_TRUE	/* set_fdfl(): set the flag */
#define	FL_UNSET	B_FALSE	/* set_fdfl(): unset the flag */

static int	nosuccess_cnt;	/* unsuccessful delivery counter */

static int	retries;		/* connection retries */
int		timeout;		/* connection timeout */
static int	timeout_p_timeout;	/* p_timeout attr storage */

/* semi-exponential timeout back off; x .. attempts, y .. timeout */
#define	BOFF_TIMEOUT(x, y)	(x < 3 ? y * 2 * x : y * 8)

/* general plugin lock */
pthread_mutex_t	plugin_mutex = PTHREAD_MUTEX_INITIALIZER;

static struct hostlist_s	*current_host;
static struct hostlist_s	*hosts;
static struct hostlist_s	*hosts_prev;

extern struct transq_hdr_s	transq_hdr;
static long			transq_count_max;
extern pthread_mutex_t		transq_lock;

extern pthread_t	recv_tid;

extern boolean_t	notify_pipe_ready;
extern int		notify_pipe[2];

#if DEBUG
FILE		*dfile;		/* debug file */
#endif

/*
 * set_transq_count_max() - sets the transq_count_max value based on kernel
 * audit queue high water mark. This is backup solution for a case, when the
 * the default qsize zero value is (intentionally) set in the audit_remote(5)
 * plugin configuration.
 */
static auditd_rc_t
set_transq_count_max()
{
	struct au_qctrl	qctrl;

	if (auditon(A_GETQCTRL, (caddr_t)&qctrl, 0) != -1) {
		transq_count_max = qctrl.aq_hiwater;
		DPRINT((dfile, "Transmission queue max length set to %ld\n",
		    transq_count_max));
		return (AUDITD_SUCCESS);
	}

	DPRINT((dfile, "Setting the transmission queue max length failed\n"));
	return (AUDITD_RETRY);
}

/*
 * get_port_default() - set the default port number; note, that "solaris-audit"
 * used below in the code is the IANA assigned service name for the secure
 * remote solaris audit logging.
 */
static auditd_rc_t
get_port_default(int *port_default)
{

	struct servent  serventry;
	char  		serventry_buf[1024];

	if (getservbyname_r("solaris-audit", "tcp", &serventry,
	    (char *)&serventry_buf, sizeof (serventry_buf)) == NULL) {
		DPRINT((dfile, "unable to get default port number\n"));
#if DEBUG
		if (errno == ERANGE) {
			DPRINT((dfile, "low on buffer\n"));
		}
#endif
		return (AUDITD_INVALID);
	}
	*port_default = ntohs(serventry.s_port);
	DPRINT((dfile, "default port: %d\n", *port_default));

	return (AUDITD_SUCCESS);
}

/*
 * trim_me() - trims the white space characters around the specified string.
 * Inputs - pointer to the beginning of the string (str_ptr); returns - pointer
 * to the trimmed string. Function returns NULL pointer in case of received
 * empty string, NULL pointer or in case the pointed string consists of white
 * space characters only.
 */
static char *
trim_me(char *str_ptr) {

	char	*str_end;

	if (str_ptr == NULL || *str_ptr == '\0') {
		return (NULL);
	}

	while (isspace(*str_ptr)) {
		str_ptr++;
	}
	if (*str_ptr == '\0') {
		return (NULL);
	}

	str_end = str_ptr + strlen(str_ptr);

	while (str_end > str_ptr && isspace(str_end[-1])) {
		str_end--;
	}
	*str_end = '\0';

	return (str_ptr);
}

/*
 * Frees host list - should be called while keeping auditd_mutex if working with
 * static hostlist_t.
 */
static void
freehostlist(hostlist_t **hostlist_ptr)
{
	hostlist_t *h, *n;

	h = *hostlist_ptr;

	while (h != NULL)  {
		n = h->next_host;
		freehostent(h->host);
		free(h);
		h = n;
	}
	*hostlist_ptr = NULL;
}

/*
 * parsehosts() end parses the host string (hosts_str)
 */
static auditd_rc_t
parsehosts(char *hosts_str, char **error)
{
	char 		*hostportmech, *hpm;
	char		*hostname;
	char		*port_str;
	char		*mech_str;
	int		port;
	int		port_default = -1;
	gss_OID		mech_oid;
	char 		*lasts_hpm;
	hostlist_t 	*lasthost = NULL;
	hostlist_t 	*hosts_new = NULL;
	hostlist_t	*newhost;
	struct hostent 	*hostentry;
	int		error_num;
	int		rc;
#if DEBUG
	char 		addr_buf[INET6_ADDRSTRLEN];
	int		num_of_hosts = 0;
#endif

	DPRINT((dfile, "parsing %s\n", hosts_str));
	while ((hostportmech = strtok_r(hosts_str, ",", &lasts_hpm)) != NULL) {

		hosts_str = NULL;
		hostname = NULL;
		port_str = NULL;
		port = port_default;
		mech_str = NULL;
		mech_oid = GSS_C_NO_OID;

		DPRINT((dfile, "parsing host:port:mech %s\n", hostportmech));

		if (strncmp(hostportmech, ":", 1 == 0)) { /* ":port:" case */
			*error = strdup(gettext("no hostname specified"));
			freehostlist(&hosts_new);
			return (AUDITD_INVALID);
		}

		/* parse single host:port:mech target */
		while ((hpm = strsep(&hostportmech, ":")) != NULL) {

			if (hostname == NULL) {
				hostname = hpm;
				continue;
			}
			if (port_str == NULL) {
				port_str = hpm;
				continue;
			}
			if (mech_str == NULL) {
				mech_str = hpm;
				continue;
			}

			/* too many colons in the hostportmech string */
			*error = strdup(gettext("invalid host:port:mech "
			    "specification"));
			freehostlist(&hosts_new);
			return (AUDITD_INVALID);
		}

		if (hostname == NULL || *hostname == '\0') {
			*error = strdup(gettext("invalid hostname "
			    "specification"));
			freehostlist(&hosts_new);
			return (AUDITD_INVALID);
		}

		/* trim hostname */
		hostname = trim_me(hostname);
		if (hostname == NULL || *hostname == '\0') {
			*error = strdup(gettext("empty hostname "
			    "specification"));
			freehostlist(&hosts_new);
			return (AUDITD_INVALID);
		}

		DPRINT((dfile, "resolving address for %s\n", hostname));

		hostentry = getipnodebyname(hostname, AF_INET6, 0, &error_num);
		if (hostentry == NULL) {
			hostentry = getipnodebyname(hostname, AF_INET, 0,
			    &error_num);
		}
		if (hostentry == NULL) {
			if (error_num == TRY_AGAIN) {
				*error = strdup(gettext("host not found, "
				    "try later"));
				freehostlist(&hosts_new);
				return (AUDITD_RETRY);
			} else {
				*error = strdup(gettext("host not found"));
				freehostlist(&hosts_new);
				return (AUDITD_INVALID);
			}
		}
		DPRINT((dfile, "hostentry: h_name=%s, addr_len=%d, addr=%s\n",
		    hostentry->h_name, hostentry->h_length,
		    inet_ntop(hostentry->h_addrtype,
		    hostentry->h_addr_list[0], addr_buf,
		    INET6_ADDRSTRLEN)));

		/* trim port */
		port_str = trim_me(port_str);
		if (port_str == NULL || *port_str == '\0') {
			if (port_default == -1 &&
			    (rc = get_port_default(&port_default))
			    != AUDITD_SUCCESS) {
				*error = strdup(gettext(
				    "unable to get default port number"));
				freehostlist(&hosts_new);
				return (rc);
			}
			port = port_default;
			DPRINT((dfile, "port: %d (default)\n", port));
		} else {
			errno = 0;
			port = atoi(port_str);
			if (errno != 0 || port < 1 || port > USHRT_MAX) {
				*error = strdup(gettext("invalid port number"));
				freehostlist(&hosts_new);
				return (AUDITD_INVALID);
			}
			DPRINT((dfile, "port: %d\n", port));
		}

		/* trim mechanism */
		mech_str = trim_me(mech_str);
		if (mech_str != NULL && *mech_str != '\0') {
			if (rpc_gss_mech_to_oid(mech_str, &mech_oid) != TRUE) {
				*error = strdup(gettext("unknown mechanism"));
				freehostlist(&hosts_new);
				return (AUDITD_INVALID);
			}
			DPRINT((dfile, "mechanism: %s\n", mech_str));
#if DEBUG
		} else {
			DPRINT((dfile, "mechanism: null (default)\n"));
#endif
		}

		/* add this host to host list */
		newhost = malloc(sizeof (hostlist_t));
		if (newhost == NULL) {
			*error = strdup(gettext("no memory"));
			freehostlist(&hosts_new);
			return (AUDITD_NO_MEMORY);
		}
		newhost->host = hostentry;
		newhost->port = htons(port);
		newhost->mech = mech_oid;
		newhost->next_host = NULL;
		if (lasthost != NULL) {
			lasthost->next_host = newhost;
			lasthost = lasthost->next_host;
		} else {
			lasthost = newhost;
			hosts_new = newhost;
		}
#if DEBUG
		num_of_hosts++;
#endif
	}

	(void) pthread_mutex_lock(&plugin_mutex);
	if (hosts_prev == NULL) {
		hosts_prev = hosts;
	}
	hosts = hosts_new;
	current_host = hosts;
	(void) pthread_mutex_unlock(&plugin_mutex);

	DPRINT((dfile, "Configured %d hosts.\n", num_of_hosts));

	return (AUDITD_SUCCESS);
}


#if DEBUG
static char *
auditd_message(auditd_rc_t msg_code) {
	char 	*rc_msg;

	switch (msg_code) {
	case AUDITD_SUCCESS:
		rc_msg = strdup("ok");
		break;
	case AUDITD_RETRY:
		rc_msg = strdup("retry after a delay");
		break;
	case AUDITD_NO_MEMORY:
		rc_msg = strdup("can't allocate memory");
		break;
	case AUDITD_INVALID:
		rc_msg = strdup("bad input");
		break;
	case AUDITD_COMM_FAIL:
		rc_msg = strdup("communications failure");
		break;
	case AUDITD_FATAL:
		rc_msg = strdup("other error");
		break;
	case AUDITD_FAIL:
		rc_msg = strdup("other non-fatal error");
		break;
	}
	return (rc_msg);
}
#endif

/*
 * rsn_to_msg() - translation of the reason of closure identifier to the more
 * human readable/understandable form.
 */
static char *
rsn_to_msg(close_rsn_t reason)
{
	char 	*rc_msg;

	switch (reason) {
	case RSN_UNDEFINED:
		rc_msg = strdup(gettext("not defined reason of failure"));
		break;
	case RSN_INIT_POLL:
		rc_msg = strdup(gettext("poll() initialization failed"));
		break;
	case RSN_TOK_RECV_FAILED:
		rc_msg = strdup(gettext("token receiving failed"));
		break;
	case RSN_TOK_TOO_BIG:
		rc_msg = strdup(gettext("unacceptable token size"));
		break;
	case RSN_TOK_UNVERIFIABLE:
		rc_msg = strdup(gettext("received unverifiable token"));
		break;
	case RSN_SOCKET_CLOSE:
		rc_msg = strdup(gettext("closed socket"));
		break;
	case RSN_SOCKET_CREATE:
		rc_msg = strdup(gettext("socket creation failed"));
		break;
	case RSN_CONNECTION_CREATE:
		rc_msg = strdup(gettext("connection creation failed"));
		break;
	case RSN_PROTOCOL_NEGOTIATE:
		rc_msg = strdup(gettext("protocol negotiation failed"));
		break;
	case RSN_GSS_CTX_ESTABLISH:
		rc_msg = strdup(gettext("context establishing failed"));
		break;
	case RSN_GSS_CTX_EXP:
		rc_msg = strdup(gettext("context expired"));
		break;
	case RSN_UNKNOWN_AF:
		rc_msg = strdup(gettext("unknown address family"));
		break;
	case RSN_MEMORY_ALLOCATE:
		rc_msg = strdup(gettext("memory allocation failed"));
		break;
	default:	/* RSN_OTHER_ERR */
		rc_msg = strdup(gettext("other, not classified error"));
		break;
	}
	return (rc_msg);
}

/*
 * set_fdfl() - based on set_fl (FL_SET/FL_UNSET) un/sets the fl flag associated
 * with fd file descriptor.
 */
static boolean_t
set_fdfl(int fd, int fl, boolean_t set_fl)
{
	int	flags;

	/* power of two test - only single bit flags are allowed */
	if (!fl || (fl & (fl-1))) {
		DPRINT((dfile, "incorrect flag - %d isn't power of two\n", fl));
		return (B_FALSE);
	}

	if ((flags = fcntl(fd, F_GETFL, 0)) < 0) {
		DPRINT((dfile, "cannot get file descriptor flags\n"));
		return (B_FALSE);
	}

	if (set_fl) {	/* set the fl flag */
		if (flags & fl) {
			return (B_TRUE);
		}

		flags |= fl;

	} else {	/* unset the fl flag */
		if (~flags & fl) {
			return (B_TRUE);
		}

		flags &= ~fl;
	}

	if (fcntl(fd, F_SETFL, flags) == -1) {
		DPRINT((dfile, "cannot %s file descriptor flags\n",
		    (set_fl ? "set" : "unset")));
		return (B_FALSE);
	}

	DPRINT((dfile, "fd: %d - flag: 0%o was %s\n", fd, fl,
	    (set_fl ? "set" : "unset")));
	return (B_TRUE);
}


/*
 * create_notify_pipe() - creates the notification pipe. Function returns
 * B_TRUE/B_FALSE on success/failure.
 */
static boolean_t
create_notify_pipe(int *notify_pipe, char **error)
{

	if (pipe(notify_pipe) < 0) {
		DPRINT((dfile, "Cannot create notify pipe: %s\n",
		    strerror(errno)));
		*error = strdup(gettext("failed to create notification pipe"));
		return (B_FALSE);
	} else {
		DPRINT((dfile, "Pipe created in:%d out:%d\n", notify_pipe[0],
		    notify_pipe[1]));
		/* make (only) the pipe "in" end nonblocking */
		if (!set_fdfl(notify_pipe[0], O_NONBLOCK, FL_UNSET) ||
		    !set_fdfl(notify_pipe[1], O_NONBLOCK, FL_SET)) {
			DPRINT((dfile, "Cannot prepare blocking scheme on top "
			    "of the notification pipe: %s\n", strerror(errno)));
			(void) close(notify_pipe[0]);
			(void) close(notify_pipe[1]);

			*error = strdup(gettext("failed to prepare blocking "
			    "scheme on top of the notification pipe"));
			return (B_FALSE);
		}
	}

	return (B_TRUE);
}


/*
 * auditd_plugin() sends a record via a tcp connection.
 *
 * Operation:
 *   - 1 tcp connection opened at a time, referenced by current_host->sockfd
 *   - tries to (open and) send a record to the current_host where its address
 *     is taken from the first hostent h_addr_list entry
 *   - if connection times out, tries second host
 *   - if all hosts where tried tries again for retries number of times
 *   - if everything fails, it bails out with AUDITD_RETRY
 *
 *   Note, that space on stack allocated for any error message returned along
 *   with AUDITD_RETRY is subsequently freed by auditd.
 *
 */
auditd_rc_t
auditd_plugin(const char *input, size_t in_len, uint64_t sequence, char **error)
{
	int 		rc = AUDITD_FAIL;
	int 		send_record_rc = SEND_RECORD_FAIL;
	hostlist_t 	*start_host;
	int 		attempts = 0;
	char		*ext_error;	/* extended error string */
	close_rsn_t	err_rsn = RSN_UNDEFINED;
	char		*rsn_msg;

#if DEBUG
	char		*rc_msg;
	static uint64_t	last_sequence = 0;

	if ((last_sequence > 0) && (sequence != last_sequence + 1)) {
		DPRINT((dfile, "audit_remote: buffer sequence=%llu "
		    "but prev=%llu\n", sequence, last_sequence));
	}
	last_sequence = sequence;

	DPRINT((dfile, "audit_remote: input seq=%llu, len=%d\n",
	    sequence, in_len));
#endif

	(void) pthread_mutex_lock(&transq_lock);

	if (transq_hdr.count == transq_count_max) {
		DPRINT((dfile, "Transmission queue is full (%ld)\n",
		    transq_hdr.count));
		(void) pthread_mutex_unlock(&transq_lock);
		*error = strdup(gettext("retransmission queue is full"));
		return (AUDITD_RETRY);
	}
	(void) pthread_mutex_unlock(&transq_lock);


	(void) pthread_mutex_lock(&plugin_mutex);

	/* cycle over the hosts and possibly deliver the record */
	start_host = current_host;
	while (rc != AUDITD_SUCCESS) {
		DPRINT((dfile, "Trying to send record to %s [attempt:%d/%d]\n",
		    current_host->host->h_name, attempts + 1, retries));

		send_record_rc = send_record(current_host, input, in_len,
		    sequence, &err_rsn);
		DPRINT((dfile, "send_record() returned %d - ", send_record_rc));

		switch (send_record_rc) {
		case SEND_RECORD_SUCCESS:
			DPRINT((dfile, "success\n"));
			nosuccess_cnt = 0;
			rc = AUDITD_SUCCESS;
			if (hosts_prev != NULL) {
				freehostlist(&hosts_prev);
				DPRINT((dfile, "stale host list freed\n"));
			}
			break;
		case SEND_RECORD_NEXT:
			DPRINT((dfile, "retry the same host: %s (penalty) "
			    "rsn:%d\n", current_host->host->h_name, err_rsn));
			attempts++;
			break;
		case SEND_RECORD_RETRY:
			DPRINT((dfile, "retry the same host: %s (no penalty) "
			    "rsn:%d\n", current_host->host->h_name, err_rsn));
			break;
		}

		if (send_record_rc == SEND_RECORD_NEXT) {

			/* warn about unsuccessful auditd record delivery */
			rsn_msg = rsn_to_msg(err_rsn);
			(void) asprintf(&ext_error,
			    "retry %d connection %s:%d %s", attempts + 1,
			    current_host->host->h_name,
			    ntohs(current_host->port), rsn_msg);
			if (ext_error == NULL) {
				free(rsn_msg);
				*error = strdup(gettext("no memory"));
				rc = AUDITD_NO_MEMORY;
				break;
			}
			__audit_dowarn2("plugin", "audit_remote.so", "retry",
			    ext_error, attempts + 1);
			free(rsn_msg);
			free(ext_error);

			if (attempts < retries) {
				/* semi-exponential timeout back off */
				timeout = BOFF_TIMEOUT(attempts, timeout);
				DPRINT((dfile, "New timeout=%d\n", timeout));
			} else {
				/* get next host */
				current_host = current_host->next_host;
				if (current_host == NULL) {
					current_host = hosts;
				}
				timeout = timeout_p_timeout;
				DPRINT((dfile, "New timeout=%d\n", timeout));
				attempts = 0;
			}

			/* one cycle finished */
			if (current_host == start_host && attempts == 0) {
				nosuccess_cnt++;
				(void) asprintf(&ext_error, "all hosts defined "
				    "as p_hosts were tried to deliver "
				    "the audit record to with no success "
				    "- sleeping for %d seconds",
				    NOSUCCESS_DELAY);
				if (ext_error == NULL) {
					*error = strdup(gettext("no memory"));
					rc = AUDITD_NO_MEMORY;
					break;
				}
				__audit_dowarn2("plugin", "audit_remote.so",
				    "retry", ext_error, nosuccess_cnt);
				free(ext_error);
				(void) sleep(NOSUCCESS_DELAY);
			}

		} /* if (send_record_rc == SEND_RECORD_NEXT) */

		err_rsn = RSN_UNDEFINED;

	} /* while (rc != AUDITD_SUCCESS) */

	(void) pthread_mutex_unlock(&plugin_mutex);

#if DEBUG
	rc_msg = auditd_message(rc);
	DPRINT((dfile, "audit_remote: returning: %s\n", rc_msg));
	free(rc_msg);
#endif

	return (rc);
}

/*
 * auditd_plugin_open() may be called multiple times; on initial open or
 * `audit -s`, then kvlist != NULL; on `audit -n`, then kvlist == NULL.
 * For more information see audit(1M).
 *
 * Note, that space on stack allocated for any error message returned along
 * with AUDITD_RETRY is subsequently freed by auditd.
 *
 */
auditd_rc_t
auditd_plugin_open(const kva_t *kvlist, char **ret_list, char **error)
{
	kva_t	*kv;
	char	*val_str;
	int	val;
	long	val_l;
	int	rc = 0;

	*error = NULL;
	*ret_list = NULL;
	kv = (kva_t *)kvlist;

#if DEBUG
	dfile = __auditd_debug_file_open();
#endif

	/* initial open or audit -s */
	if (kvlist != NULL) {
		DPRINT((dfile, "Action: initial open or `audit -s`\n"));
		val_str = kva_match(kv, "p_timeout");
		if (val_str == NULL) {
			*error = strdup(
			    gettext("p_timeout attribute not found"));
			return (AUDITD_RETRY);
		}
		DPRINT((dfile, "val_str=%s\n", val_str));
		errno = 0;
		val = atoi(val_str);
		if (errno == 0 && val >= 1) {
			timeout_p_timeout = val;
			timeout = val;
		} else {
			timeout_p_timeout = DEFAULT_TIMEOUT;
			timeout = timeout_p_timeout;
			DPRINT((dfile, "p_timeout set to default value: %d\n",
			    timeout));
		}

		val_str = kva_match(kv, "p_retries");
		if (val_str == NULL) {
			*error = strdup(
			    gettext("p_retries attribute not found"));
			return (AUDITD_RETRY);
		}
		DPRINT((dfile, "val_str=%s\n", val_str));
		errno = 0;
		val = atoi(val_str);
		if (errno == 0 && val >= 0) {
			retries = val;
		}

		val_str = kva_match(kv, "qsize");
		if (val_str == NULL) {
			*error = strdup(gettext("qsize attribute not found"));
			return (AUDITD_RETRY);
		}
		DPRINT((dfile, "qsize=%s\n", val_str));
		errno = 0;
		val_l = atol(val_str);
		if (errno == 0 && val_l >= 0) {
			transq_count_max = val_l;
		}
		if (transq_count_max == 0 &&
		    (rc = set_transq_count_max()) != AUDITD_SUCCESS) {
			*error = strdup(gettext("cannot get kernel "
			    "auditd queue high water mark\n"));
			return (rc);
		}
		DPRINT((dfile, "timeout=%d, retries=%d, transq_count_max=%ld\n",
		    timeout, retries, transq_count_max));

		val_str = kva_match(kv, "p_hosts");
		if (val_str == NULL) {
			*error = strdup(gettext("no hosts configured"));
			return (AUDITD_RETRY);
		}
		if ((rc = parsehosts(val_str, error)) != AUDITD_SUCCESS) {
			return (rc);
		}

		/* create the notification pipe towards the receiving thread */
		if (!notify_pipe_ready) {
			if (create_notify_pipe(notify_pipe, error)) {
				notify_pipe_ready = B_TRUE;
			} else {
				return (AUDITD_RETRY);
			}
		}

#if DEBUG
	} else { /* audit -n */
		DPRINT((dfile, "Action: `audit -n`\n"));
#endif
	}

	return (AUDITD_SUCCESS);
}

/*
 * auditd_plugin_close() performs shutdown operations. The return values are
 * used by auditd to output warnings via the audit_warn(1M) script and the
 * string returned via "error_text", is passed to audit_warn.
 *
 * Note, that space on stack allocated for any error message returned along
 * with AUDITD_RETRY is subsequently freed by auditd.
 *
 */
auditd_rc_t
auditd_plugin_close(char **error)
{
	reset_transport(DO_EXIT, DO_SYNC);
	if (pthread_join(recv_tid, NULL) != 0) {
		*error = strdup(gettext("unable to close receiving thread"));
		return (AUDITD_RETRY);
	}

	(void) pthread_mutex_lock(&plugin_mutex);
	freehostlist(&hosts);
	freehostlist(&hosts_prev);
	(void) pthread_mutex_unlock(&plugin_mutex);
	current_host = NULL;
	*error = NULL;
	return (AUDITD_SUCCESS);
}
